### ISSUE TITLE
Connection.send() never returns with the asyncio client

### ISSUE THREAD

**Issue by parmentelat** [python-websockets/websockets#1563]
hiya there

some background for starters
recently as part of an upgrade to fedora41 I have come across release 14 of `websockets`
so I have taken that chance to upgrade my code to use this new API 
turns out it is used in a monitoring process, that acts as a ws client to push its results upstream; and I am now facing a nasty sneaky issue with that, because after some unpredicatable amount of time - typically hours - the monitor stops looking at some nodes in a seemingly random way; that is to say, its focus shrinks overtime, and it reports the status of fewer and fewer nodes.

admittedly there have been many moving pieces since last time where this code was working fine
however after a week patiently adding debugging instructions, I have been able to nail it down a bit
and I have evidence that in my application, it is the call to
`websockets.asyncio.client.ClientConnection.send()`
that essentially never returns; like in, never ever even after hours
specifically I am using 14.1, and the `websockets.asyncio.client` flavour

so, that's all I have as hard evidence for now
I am still miles away from being able to reproduce on a simple example - I can't even do that reliably in my complex app, I only know to wait for the bug to manifest itself...

at this point I have opted for rolling back to using `websockets.legacy` instead, still with 14.1
and so far this has been performing all right for a full day, so I'm optimistic my issue is fixed with that older API

other than that:
as far as I can tell - but that's probably not that far - the first suspect could be this pair of lines here

https://github.com/python-websockets/websockets/blob/0403823185b272ae389da1c9182773932b4df950/src/websockets/asyncio/connection.py#L451-L452

but of course I don't know that for sure, I reckon this is just a wild guess

so, that's where I am at right now, I thought this was still worth reporting, so here I am

thanks for any hint or insight wrt to this situation

**Comment by aaugustin** [comment]
Oh, how annoying :-(

Let's rule out unlikely causes:

1. You aren't sending fragmented messages i.e. calling `send()` with an iterable or async iterable, are you?
2. You kept `close_timeout` set to the default value or set a reasonable value, didn't you?
3. You aren't sending large enough amounts of data to create a risk of clogging the correction, are you?

Then, if I had to bet:

1. `async with self.send_context(): ...` could be blocking during the exit of the context manager. In that case, when sending a text message, [this line](https://github.com/python-websockets/websockets/blob/d852df7dd6324eaee17fc848f029ada371678cbe/src/websockets/asyncio/connection.py#L462) would execute but you'd never exit the context manager and jump to the next line.
2. You could be losing the TCP connection in a way that websockets doesn't handle well. In order to prevent leaking TCP connection, websockets waits for the TCP connection to be lost before returning. It took a long time to iron out all potential issues in the original implementation; I wouldn't be surprised if the rewrite had some bugs; the I/O layer is brand new and more difficult to get right than the Sans-I/O layer.

**Comment by aaugustin** [comment]
Thank you for reporting the issue BTW.

**Comment by parmentelat** [comment]
hiya there again

many thanks for the quick feedback :)

none of the 3 unlikely causes seem to apply to me; I pass `send()` a good-old `str` object; I don't mess with the timeouts; my chunks are pretty small, in the order of 1k chars or similar
between the 2 other I coul not say which is the most likely in my case; it *feels* however like my issues *might* be somehow related to the server end being restarted in some way, I have no hard evidence about that but it's possible

I'm unsure how to go about helping further, insofar as I have kind of several times overburnt my anticipated time budget on this topic :) 
the service based on `legacy` has been running for 36 hours now without a glitch, so I guess for now I'm going to stay put and leave things as they are

let me know if you do any potential progress around this issue, I will try to make it simple for me to try out any future version down the line

thanks again for the insights, much appreciated !

**Comment by aaugustin** [comment]
Thank you for the information you provided. I'm going to look into it. Probably this will get ironed out long before the deprecation of the legacy implementation completes (5 years). I'm sorry for the rocky migration experience.

**Comment by parmentelat** [comment]
thanks for the heads up - I'm of course still interested in any further progress :)

**Comment by Electron4444** [comment]
I'm not sure if this is part of the problem, but I noticed that in the legacy version, `WebSocketCommonProtocol.send` first checks whether the connection state is still open before performing type matching, whereas `Connection.send` does not (could check `self.protocol.state`, right?). That said, I might be completely wrong here, so feel free to disregard this if I’ve missed some underlying implementation detail.

**Comment by aaugustin** [comment]
Opus has a full theory supporting that it's a duplicate of #1555 (caveat: sycophancy, etc.)

I'm going to call it fixed in 14.2. You're welcome to try again with the latest release 🤞 

<details>
Root cause: send_context() can wait forever on connection_lost_waiter

You're right that fragmented_send_waiter is a red herring. Every send() — fragmented or not — funnels through send_context() (connection.py:834), and the hang lives in its error/close-exit path.

The stuck code (14.1, connection.py:927‑933)

if raise_close_exc:
    self.close_transport()                              # -> self.transport.close()
    await asyncio.shield(self.connection_lost_waiter)   # <-- UNBOUNDED wait
    raise self.protocol.close_exc from original_exc

close_transport() calls transport.close(), then the code waits on connection_lost_waiter with no timeout. That future is only resolved by connection_lost() (connection.py:1011). So if connection_lost() never fires, send() never returns.

Why transport.close() fails to trigger connection_lost()

CPython's _SelectorTransport.close() only schedules connection_lost() immediately if the write buffer is empty. If data is still buffered, it defers until the buffer drains via the socket becoming writable. Two conditions, occurring together, make that drain never happen:

1. Network unavailable / peer not reading → the socket never becomes writable, so the pending write buffer never flushes. Normally the read side would still detect a reset/EOF and close the transport, but…
2. Receive buffer saturated → when the app doesn't consume recv() fast enough, the Assembler crosses its high-water mark and calls transport.pause_reading() (messages.py:260‑269, wired up at connection.py:985‑989). With the reader removed, the transport can't detect the dead connection from the read side either.

Write side can't flush, read side is disabled → the transport is wedged → connection_lost() is never called → connection_lost_waiter never resolves → send() hangs forever.

Note the close_deadline/close_timeout guard on the earlier await (connection.py:913‑925) doesn't save you: on timeout it just sets raise_close_exc = True and falls through into the unguarded await above.

Why this matches the report exactly

A monitoring client fanned out across many nodes is the perfect trigger: it doesn't always drain recv() promptly (→ pause_reading), and over hours individual nodes hit network trouble (→ stalled writes). Each time both line up on a connection, that connection's send() wedges permanently — "the number of monitored nodes decreases over time." The legacy API uses a different I/O path, which is why rolling back worked.

The fix (already shipped in 14.2, commit 613f3f0)

-            self.close_transport()
+            self.transport.abort()

transport.abort() drops the buffers and forces connection_lost() to be called promptly, so the waiter resolves and send() raises ConnectionClosed instead of hanging. close_transport() was removed.

One thing worth flagging: that commit is labeled "Fix #1555 (for real!)" and the 14.2 changelog phrases it as a close() fix — but since send(), close(), ping(), and keepalive() all share send_context(), the same commit fixes #1563 too. #1563 is effectively a duplicate of #1555 with the symptom surfacing on send() rather than close(). If it's still open, it can be closed as fixed in 14.2.
</details>
